src/Namespace.sol

```diff
@@ -0,0 +1,578 @@
+        1924992000,
+        1956528000,
+        1988150400,
+        2019686400,
+        2051222400,
+        2082758400,
+        2114380800,
+        2145916800
+    ];
+
+    address public immutable vault;
+
+    uint256 immutable escrowPeriod = 3 days;
+
+    constructor(
+        string memory _name,
+        string memory _symbol,
+        address _owner,
+        address _vault
+    ) ERC721(_name, _symbol) Owned(_owner) {
+        vault = _vault;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                             REGISTRATION LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * INVARIANT 1A: If an id is not minted, expiryOf[id] must be 0 and  _ownerOf[id] and
+     *               recoveryOf[id] must also be address(0).
+     *
+     * INVARIANT 1B: If an id is minted, expiryOf[id] and _ownerOf[id] must be non-zero.
+     *
+     * INVARIANT 2: A username cannot be transferred to address(0) after it is minted.
+     */
+
+    /**
+     * @notice Generate a commitment hash used to secretly lock down a username for registration.
+     *
+     * @dev The commitment process prevents front-running of the username registration.
+     *
+     * @param username the username to be registered
+     * @param owner the address that will own the username
+     * @param secret a salt that randomizes and secures the commitment hash
+     */
+    function generateCommit(
+        bytes16 username,
+        address owner,
+        bytes32 secret
+    ) public pure returns (bytes32) {
+        if (!_isValidUsername(username)) revert InvalidName();
+
+        return keccak256(abi.encode(username, owner, secret));
+    }
+
+    /**
+     * @notice Save a generated commitment hash on-chain, which must be done before a registration
+     *         can occur
+     *
+     * @dev The commitment process prevents front-running of the username registration.
+     *
+     * @param commit the commitment hash to be persisted on-chain
+     */
+    function makeCommit(bytes32 commit) public {
+        stateOf[commit] = true;
+    }
+
+    /**
+     * @notice Mint a new username if a commitment was made previously and send it to the owner.
+     *
+     * @param username the username to register
+     * @param owner the address that will claim the username
+     * @param secret the secret that protects the commitment
+     */
+    function register(
+        bytes16 username,
+        address owner,
+        bytes32 secret
+    ) external payable {
+        bytes32 commit = generateCommit(username, owner, secret);
+
+        uint256 _currYearFee = currYearFee();
+        if (msg.value < _currYearFee) revert InsufficientFunds();
+
+        if (!stateOf[commit]) revert InvalidCommit();
+        delete stateOf[commit];
+
+        uint256 tokenId = uint256(bytes32(username));
+        if (expiryOf[tokenId] != 0) revert NotRegisterable();
+
+        _mint(owner, tokenId);
+
```

Comment:
Should we emit an event here for registration? (or are we assuming the _mint one is enough)